notanorm/base.py

```diff
@@ -99,17 +115,18 @@ class DbBase(ABC):                          # pylint: disable=too-many-public-me
     reconnect_backoff_factor = 2
     debug_sql = None
     debug_args = None
+    __lock_pool = defaultdict(threading.RLock)
 
     @property
     def timeout(self):
         # total timeout for connections
-        return self.max_reconnect_attempts * (self.reconnect_backoff_start ** self.reconnect_backoff_factor)
+        return self.reconnect_backoff_start * (self.max_reconnect_attempts ** self.reconnect_backoff_factor)
 
     def __init__(self, *args, **kws):
         self.__conn_p = None
-        self.__conn_args = args
-        self.__conn_kws = kws
-        self.r_lock = threading.RLock()
+        self._conn_args = args
+        self._conn_kws = kws
+        self.r_lock = self.__lock_pool[(args, tuple(sorted((k, v) for k, v in kws.items())))]
```

Comment:
this seems fishy... maybe the subclass should just define what this key should be? Otherwise we e.g. have different locks for different usernames and the same db, in mysql

Reply:
yep, and really mysql probably doesn't need or want this lock.   i will check it.